WordPress/Classes/Services/MediaImportService.swift

```diff
@@ -20,55 +20,254 @@ open class MediaImportService: LocalCoreDataService {
+    ///         localized description. This closure is called on the main thread.
     ///
     /// - Returns: a progress object that report the current state of the import process.
     ///
-    @objc(importResource:toMedia:onCompletion:onError:)
-    func `import`(_ exportable: ExportableAsset, to media: Media, onCompletion: @escaping MediaCompletion, onError: @escaping OnError) -> Progress? {
+    private func `import`(_ exportable: ExportableAsset, to media: Media, allowableFileExtensions: Set<String>, completion: @escaping (Result<Media, Error>) -> Void) -> Progress {
         let progress: Progress = Progress.discreteProgress(totalUnitCount: 1)
         importQueue.async {
-            guard let exporter = self.makeExporter(for: exportable) else {
+            guard let exporter = self.makeExporter(for: exportable, allowableFileExtensions: allowableFileExtensions) else {
                 preconditionFailure("An exporter needs to be availale")
             }
-            let exportProgress = exporter.export(onCompletion: { export in
-                self.managedObjectContext.perform {
-                    self.configureMedia(media, withExport: export)
-                    ContextManager.sharedInstance().save(self.managedObjectContext, completion: {
-                        onCompletion(media)
+            let exportProgress = exporter.export(
+                onCompletion: { export in
+                    self.coreDataStack.performAndSave({ context in
+                        let mediaInContext = try context.existingObject(with: media.objectID) as! Media
+                        self.configureMedia(mediaInContext, withExport: export)
+                    }, completion: { result in
+                        completion(
+                            result.flatMap {
+                                Result {
+                                    try self.coreDataStack.mainContext.existingObject(with: media.objectID) as! Media
+                                }
+                            }
+                        )
                     }, on: .main)
+                },
+                onError: { error in
+                    MediaImportService.logExportError(error)
+                    // Return the error via the context's queue, and as an NSError to ensure it carries over the right code/message.
+                    DispatchQueue.main.async {
+                        completion(.failure(error))
+                    }
                 }
-            }, onError: { mediaExportError in
-                self.handleExportError(mediaExportError, errorHandler: onError)
-            }
             )
             progress.addChild(exportProgress, withPendingUnitCount: 1)
         }
         return progress
     }
 
-    func makeExporter(for exportable: ExportableAsset) -> MediaExporter? {
+    private func makeExporter(for exportable: ExportableAsset, allowableFileExtensions: Set<String>) -> MediaExporter? {
```

Comment:
Have you considered using "allowed"?

**Update:** Oh, I see that name was already used. 🤔 No point in using a different one, then, it would end up being more confusing than useful, even if we were to agree that "allowed" is in itself clearer than "allowable".